<h2>Práctica Tema 2 </h2>
<h3>Módulo 3 - Manipulación de Datos en Python </h3>

Descarga el siguiente archivo XLS :
https://file-examples-com.github.io/uploads/2017/02/file_example_XLS_5000.xls

Para tu mayor comodidad, este archivo ya ha sido descargado y ya está preparado en esta carpeta

1. El primer paso es carga el archivo en un DataFrame

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
dataframe = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/Data Science and AI/Tercera sesión/Ejemplo1.xls')

Mounted at /content/drive


In [2]:
dataframe

Unnamed: 0 First Name   Last Name  Gender        Country  Age  \
0              1      Dulce       Abril  Female  United States   32   
1              2       Mara   Hashimoto  Female  Great Britain   25   
2              3     Philip        Gent    Male         France   36   
3              4   Kathleen      Hanner  Female  United States   25   
4              5    Nereida     Magwood  Female  United States   58   
...          ...        ...         ...     ...            ...  ...   
4995        4996       Roma  Lafollette  Female  United States   34   
4996        4997     Felisa        Cail  Female  United States   28   
4997        4998   Demetria       Abbey  Female  United States   32   
4998        4999     Jeromy        Danz    Male  United States   39   
4999        5000   Rasheeda      Alkire  Female  United States   29   

            Date    Id  
0     15/10/2017  1562  
1     16/08/2016  1582  
2     21/05/2015  2587  
3     15/10/2017  3549  
4     16/08/2016  2468  
...          ...   ...  
4995  15/10/2017  2654  
4996  16/08/2016  6525  
4997  21/05/2015  3265  
4998  15/10/2017  3265  
4999  16/08/2016  6125  

[5000 rows x 8 columns]

2. El siguente paso es eliminar todos los registros que sean de Estados Unidos

In [3]:
dataframe = dataframe[dataframe['Country'] != 'United States']
dataframe

Unnamed: 0 First Name  Last Name  Gender        Country  Age  \
1              2       Mara  Hashimoto  Female  Great Britain   25   
2              3     Philip       Gent    Male         France   36   
6              7       Etta       Hurn  Female  Great Britain   56   
9             10     Fallon    Winward  Female  Great Britain   28   
10            11    Arcelia     Bouska  Female  Great Britain   39   
...          ...        ...        ...     ...            ...  ...   
4981        4982     Stasia     Becker  Female  Great Britain   34   
4982        4983       Jona    Grindle  Female  Great Britain   26   
4983        4984      Judie   Claywell  Female         France   35   
4986        4987     Kelsie    Wachtel  Female         France   27   
4992        4993      Angel      Sanor    Male         France   24   

            Date    Id  
1     16/08/2016  1582  
2     21/05/2015  2587  
6     15/10/2017  3598  
9     16/08/2016  5486  
10    21/05/2015  1258  
...          ...   ...  
4981  16/08/2016  7521  
4982  21/05/2015  6458  
4983  16/08/2016  7569  
4986  16/08/2016  8642  
4992  15/10/2017  3259  

[2600 rows x 8 columns]

3. Luego, Muestra el número de personas registradas de cada país (ya sin los de Estados Unidos). Se espera una tabla que diga país y número de personas.

In [4]:
dataframe['Country'].groupby(dataframe['Country']).count().rename(index = {'Country': 'País'}).rename("Número de personas")

Country
France           1200
Great Britain    1400
Name: Número de personas, dtype: int64

4. Ahora, actualiza el nombre de los países para que el nombre esté en español

In [5]:
dataframe.loc[(dataframe['Country'] == "France"), 'Country'] = "Francia"
dataframe.loc[(dataframe['Country'] == "Great Britain"), 'Country'] = "Gran Bretaña"
dataframe

Unnamed: 0 First Name  Last Name  Gender       Country  Age        Date  \
1              2       Mara  Hashimoto  Female  Gran Bretaña   25  16/08/2016   
2              3     Philip       Gent    Male       Francia   36  21/05/2015   
6              7       Etta       Hurn  Female  Gran Bretaña   56  15/10/2017   
9             10     Fallon    Winward  Female  Gran Bretaña   28  16/08/2016   
10            11    Arcelia     Bouska  Female  Gran Bretaña   39  21/05/2015   
...          ...        ...        ...     ...           ...  ...         ...   
4981        4982     Stasia     Becker  Female  Gran Bretaña   34  16/08/2016   
4982        4983       Jona    Grindle  Female  Gran Bretaña   26  21/05/2015   
4983        4984      Judie   Claywell  Female       Francia   35  16/08/2016   
4986        4987     Kelsie    Wachtel  Female       Francia   27  16/08/2016   
4992        4993      Angel      Sanor    Male       Francia   24  15/10/2017   

        Id  
1     1582  
2     2587  
6     3598  
9     5486  
10    1258  
...    ...  
4981  7521  
4982  6458  
4983  7569  
4986  8642  
4992  3259  

[2600 rows x 8 columns]

5. el siguiente paso es cambiar el id de <b>Judie Claywell Female France</b> de 7569 a 99999

In [6]:
dataframe.loc[(dataframe['First Name'] == "Judie") & (dataframe['Last Name'] == "Claywell"), "Id"] = 99999
dataframe

Unnamed: 0 First Name  Last Name  Gender       Country  Age        Date  \
1              2       Mara  Hashimoto  Female  Gran Bretaña   25  16/08/2016   
2              3     Philip       Gent    Male       Francia   36  21/05/2015   
6              7       Etta       Hurn  Female  Gran Bretaña   56  15/10/2017   
9             10     Fallon    Winward  Female  Gran Bretaña   28  16/08/2016   
10            11    Arcelia     Bouska  Female  Gran Bretaña   39  21/05/2015   
...          ...        ...        ...     ...           ...  ...         ...   
4981        4982     Stasia     Becker  Female  Gran Bretaña   34  16/08/2016   
4982        4983       Jona    Grindle  Female  Gran Bretaña   26  21/05/2015   
4983        4984      Judie   Claywell  Female       Francia   35  16/08/2016   
4986        4987     Kelsie    Wachtel  Female       Francia   27  16/08/2016   
4992        4993      Angel      Sanor    Male       Francia   24  15/10/2017   

         Id  
1      1582  
2      2587  
6      3598  
9      5486  
10     1258  
...     ...  
4981   7521  
4982   6458  
4983  99999  
4986   8642  
4992   3259  

[2600 rows x 8 columns]

6. Ahora, crea un archivo de Excel en tu drive, en la misma carpeta en la que estas trabajando, que contenga exactamente el mismo primer renglón del archivo descargado(el de excel), pero ahora Inserta 3 registros con gente de México con ids de 5 dígitos que no inicien en cero

In [7]:
dataframe2 = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/Data Science and AI/Tercera sesión/Ejemplo3.xlsx')
dataframe2

Unnamed: 0 First Name Last Name  Gender Country  Age       Date     Id
0           1      Cielo    Garcia  Female  México   25 1999-04-09  14600
1           2     Elvira  Santiago  Female  México   65 1961-02-14  12589
2           3       Yael     Ramos    Male  México   26 2015-05-21  13245
3           4      Heidi  Gallegos  Female  México   42 2017-10-15  98367
4           5     Amaury     Felix    Male  México   24 2016-08-16  45286

7. Carga este segundo archivo como un dataFrame y concaténalo a los datos del archivo descargado

In [8]:
dataframe = pd.concat([dataframe, dataframe2])
dataframe

/usr/local/lib/python3.11/dist-packages/google/colab/_dataframe_summarizer.py:88: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  cast_date_col = pd.to_datetime(column, errors="coerce")


Unnamed: 0 First Name  Last Name  Gender       Country  Age  \
1            2       Mara  Hashimoto  Female  Gran Bretaña   25   
2            3     Philip       Gent    Male       Francia   36   
6            7       Etta       Hurn  Female  Gran Bretaña   56   
9           10     Fallon    Winward  Female  Gran Bretaña   28   
10          11    Arcelia     Bouska  Female  Gran Bretaña   39   
..         ...        ...        ...     ...           ...  ...   
0            1      Cielo     Garcia  Female        México   25   
1            2     Elvira   Santiago  Female        México   65   
2            3       Yael      Ramos    Male        México   26   
3            4      Heidi   Gallegos  Female        México   42   
4            5     Amaury      Felix    Male        México   24   

                   Date     Id  
1            16/08/2016   1582  
2            21/05/2015   2587  
6            15/10/2017   3598  
9            16/08/2016   5486  
10           21/05/2015   1258  
..                  ...    ...  
0   1999-04-09 00:00:00  14600  
1   1961-02-14 00:00:00  12589  
2   2015-05-21 00:00:00  13245  
3   2017-10-15 00:00:00  98367  
4   2016-08-16 00:00:00  45286  

[2605 rows x 8 columns]

8. Vuelve a mostrar el número de personas registradas de cada país

In [12]:
gentePorPais = dataframe['Country'].groupby(dataframe['Country']).count().rename(index = {'Country': 'País'}).rename("Número de personas")
gentePorPais

Country
Francia         1200
Gran Bretaña    1400
México             5
Name: Número de personas, dtype: int64

In [10]:
# Generar el archivo en excel
!pip install xlsxwriter
import xlsxwriter

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 165.1/165.1 kB 4.7 MB/s eta 0:00:00


In [14]:
gentePorPais.to_excel('/content/drive/MyDrive/Colab Notebooks/Data Science and AI/Tercera sesión/gentePorPais.xlsx', sheet_name = "Gente por país", engine = "xlsxwriter")